File: 0131-palindrome-partitioning/javascript/Solution.js
/**
 * @param {string} s
 * @return {string[][]}
 */
var partition = function(s) {
    let result = []
    let path = []
    let backtrack = k => {
        if (k == s.length) {
            result.push([...path])
            return
        }
        for (let end = k; end < s.length; ++end) {
            if (ispalindrome(s, k, end)) {
                path.push(s.substring(k, end+1))
                backtrack(end+1)
                path.pop()
            }
        }
    }
    backtrack(0)
    return result
};
var ispalindrome = (s, p, r) => {
    let i = p
    let j = r
    while (i <= j) {
        if (s[i] != s[j]) {
            return false
        }
        i++
        j--
    }
    return true
}

File: 0131-palindrome-partitioning/python/Solution.py
class Solution:
    def __init__(self):
        self.result = []

    def partition(self, s: str) -> List[List[str]]:
        self.backtrack(s, 0, [])
        return self.result

    def backtrack(self, s, k, path):
        if k == len(s):
            self.result.append(path[:])
            return
        for end in range(k, len(s)):
            if self.ispalindrome(s, k, end):
                path.append(s[k:end+1])
                self.backtrack(s, end+1, path)
                path.pop()

    def ispalindrome(self, s, p, r):
        i = p
        j = r
        while i <= j:
            if s[i] !=s[j]:
                return False
            i += 1
            j -= 1
        return True
